feat: blocking provider mutator

Kavindu-Dodan opened the pull request:
## This PR

- Add `SetProviderAndWait` & `SetNamedProviderAndWait` with blocking provider mutators
- Improve non-eventing provider registration by emitting ready event : This can be considered as a fix as this is defined in spec but was missing in the implementation [^1]

> The provider MAY define a status field/accessor which indicates the readiness of the provider, with possible values NOT_READY, READY, STALE, or ERROR.

> Providers without this field can be assumed to be ready immediately.

### Related Issues

Fixes #240

### How to test

Tests were added to cover new APIs and old tests were updated to handle readiness emitted for non eventing providers 

[^1]: https://openfeature.dev/specification/sections/providers#requirement-242

thisthat commented on openfeature/api.go:
```diff
@@ -211,6 +193,31 @@ func (api *evaluationAPI) initNewAndShutdownOld(newProvider FeatureProvider, old
+	err := handler.Init(apiCtx)
+	if err != nil {
+		event.EventType = ProviderError
+		event.Message = fmt.Sprintf("Provider initialization error, %v", err)
```
I think we should also clean up the `ProviderEventDetails` property otherwise it will be quite weird to fail successfully 🤣

toddbaert replied:
![image](https://github.com/open-feature/go-sdk/assets/25272906/7642b0b8-e021-409b-baad-d59dffd2fc8c)

Kavindu-Dodan replied:
@thisthat  good concern & @toddbaert  nice one 😅 

Luckily in OF go-sdk, this is handled correctly as we use go struct embedding [1] where `message` updated in error is from `ProviderEventDetails`

![image](https://github.com/open-feature/go-sdk/assets/8186721/0a17ad30-a379-4d7a-852d-954cd3315e10)

I know this is little confusing but I wanted to avoid multiple event initializations and went ahead with this. Let me know if you think we need to change this :) 

[1] - https://github.com/open-feature/go-sdk/blob/main/openfeature/provider.go#L105-L109

thisthat replied:
oohh gotcha! All good for me then 👍

toddbaert replied:
I always forget about struct embedding!

toddbaert commented on openfeature/api.go:
```diff
@@ -28,28 +29,27 @@ func newEvaluationAPI() evaluationAPI {
-// setProvider sets the default FeatureProvider of the evaluationAPI. Returns an error if FeatureProvider is nil
-func (api *evaluationAPI) setProvider(provider FeatureProvider) error {
+// setProvider sets the default FeatureProvider of the evaluationAPI.
+// Returns an error if provider registration cause an error
```
```suggestion
// Returns an error if provider registration or initialization fails
```

toddbaert commented on openfeature/openfeature.go:
```diff
@@ -20,13 +20,23 @@ func initSingleton() {
+
+// SetProviderAndWait sets the default provider and waits for its initialization.
+func SetProviderAndWait(provider FeatureProvider) error {
+	return api.setProvider(provider, false)
```
This function needs to return an error if init fails (and I think you should add documentation saying so)

https://openfeature.dev/specification/sections/flag-evaluation/#requirement-1124

Kavindu-Dodan replied:
Good catch, I will improve this

Kavindu-Dodan replied:
Done - https://github.com/open-feature/go-sdk/pull/251/commits/54ef5b0457b26309d3d4215eac8abc148de1a615

toddbaert reviewed (changes requested):
This seems good, code is clean. I think my only issue is that the set+wait function should return errors if the provider init fails, that's a key difference between the blocking and non-blocking setProvider functions.

Returning an error allows users to completely sidestep setting up error handers (though they can still do so optionally and we'd expect them to fire).

I think you should implement this and add a couple tests:

- one that returns an error for an init after some delay
- one that returns an error and also ensures error handlers are run.

Kavindu-Dodan commented:
> This seems good, code is clean. I think my only issue is that the set+wait function should return errors if the provider init fails, that's a key difference between the blocking and non-blocking setProvider functions.
> 
> Returning an error allows users to completely sidestep setting up error handers (though they can still do so optionally and we'd expect them to fire).
> 
> I think you should implement this and add a couple tests:
> 
> * one that returns an error for an init after some delay
> * one that returns an error and also ensures error handlers are run.
> 
> Relavant PR language improvement (editorial, not significant): [open-feature/spec#242](https://github.com/open-feature/spec/pull/242)

Done, addressed this concern in the latest commit. I added error return as well as test to validate error return as well as eventing

toddbaert commented on openfeature/openfeature_test.go:
```diff
@@ -252,6 +253,146 @@ func TestRequirement_1_1_2_3(t *testing.T) {
+		case errEvent = <-errChan:
+			break
+		}
+
+		if errEvent.Message == "" {
+			t.Fatal("expected non empty event message, but got empty")
+		}
+
+	})
+
+	t.Run("async registration to validate by contradiction", func(t *testing.T) {
+		// given - a provider with state handling capability, with substantial initializing delay
+		var initialized = false
+
+		provider := struct {
+			FeatureProvider
+			StateHandler
+		}{
+			NoopProvider{},
+			&stateHandlerForTests{
+				initF: func(e EvaluationContext) error {
+					<-time.After(200 * time.Millisecond)
+					initialized = true
+					return nil
+				},
+			},
+		}
+
+		// when - registered async
+		err := SetProvider(provider)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		// then - must not update variable as registration is async
+		if initialized != false {
+			t.Errorf("expected uninitialized as async, but got true")
+		}
+	})
+}
```
beautiful!
